source/extensions/access_loggers/common/grpc_access_logger.h

```diff
@@ -0,0 +1,231 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/local_info/local_info.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/stats/scope.h"
+#include "envoy/thread_local/thread_local.h"
+
+#include "common/common/assert.h"
+#include "common/grpc/typed_async_client.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Common {
+
+enum class GrpcAccessLoggerType { TCP, HTTP };
+
+namespace Detail {
+
+/**
+ * Fully specialized types of the interfaces below are available through the
+ * `Common::GrpcAccessLogger::Interface` and `Common::GrpcAccessLoggerCache::interface`
+ * aliases.
+ */
+
+/**
+ * Interface for an access logger. The logger provides abstraction on top of gRPC stream, deals with
+ * reconnects and performs batching.
+ */
+template <typename HttpLogProto, typename TcpLogProto> class GrpcAccessLogger {
```

Comment:
Does it make sense to also split the TCP and HTTP logging?
There's nothing specific to HTTP/TCP in this interface and it would allow creating a logger that does only one of the two.

Reply:
We can not split it as such but we can provide partial specialization for the `GrpcAccessLogger` that has only one `log` method, instead of two. There is currently no use case for it now, but it can be added as part of adding OLTP logger if it only supports logging one type of log entry.

```C++
template <typename LogProto> class GrpcAccessLogger<LogProto, void> {
public:
  using SharedPtr = std::shared_ptr<GrpcAccessLogger>;

  virtual ~GrpcAccessLogger() = default;

  /**
   * Log access entry.
   * @param entry supplies the access log to send.
   */
  virtual void log(LogProto&& entry) PURE;
};
```